modular_skyrat/modules/reagent_forging/code/forge.dm

```diff
@@ -80,213 +139,233 @@
 			to_chat(user, span_warning("You cannot do multiple things at the same time!"))
 			return
 		in_use = TRUE
-		if(forge_fuel_strong >= 300) //cannot insert too much
-			to_chat(user, span_warning("You only need one to two pieces of coal at a time! You have [forge_fuel_strong] seconds remaining!"))
-			in_use = FALSE
+		if(forge_fuel_strong >= 5 MINUTES) //cannot insert too much
+			fail_message(user, "You only need one piece of coal at a time! You have [forge_fuel_strong/10] seconds remaining!")
 			return
-		to_chat(user, span_warning("You start to throw the fuel into the forge..."))
-		if(!do_after(user, 3 SECONDS, target = src)) //need 3 seconds to fuel the forge
-			to_chat(user, span_warning("You abandon fueling the forge."))
-			in_use = FALSE
+		to_chat(user, span_warning("You start to throw [I] into [src]..."))
+		if(!do_after(user, skill_modifier, target = src)) //need 3 seconds to fuel the forge
+			fail_message(user, "You abandon fueling [src].")
 			return
 		var/obj/item/stack/sheet/stack_sheet = I
 		if(!stack_sheet.use(1)) //need to be able to use the item, so no glue
-			to_chat(user, span_warning("You abandon fueling the forge."))
-			in_use = FALSE
+			fail_message(user, "You abandon fueling [src].")
 			return
-		forge_fuel_strong += 300 //5 minutes
+		forge_fuel_strong += 5 MINUTES
 		in_use = FALSE
-		to_chat(user, span_notice("You successfully fuel the forge."))
+		to_chat(user, span_notice("You successfully fuel [src]."))
+		user.mind.adjust_experience(/datum/skill/smithing, 6) //useful fueling means you get some experience
 		return
 
 	if(istype(I, /obj/item/forging/billow))
 		var/obj/item/forging/forge_item = I
+		var/skill_modifier = user.mind.get_skill_modifier(/datum/skill/smithing, SKILL_SPEED_MODIFIER) * forge_item.work_time
 		if(in_use) //no spamming the billows
 			to_chat(user, span_warning("You cannot do multiple things at the same time!"))
 			return
 		in_use = TRUE
 		if(!forge_fuel_strong && !forge_fuel_weak) //if there isnt any fuel, no billow use
-			to_chat(user, span_warning("You cannot use the billow without some sort of fuel in the forge!"))
-			in_use = FALSE
-			return
-		if(forge_temperature >= 100) //we don't want the "temp" to overflow or something somehow
-			to_chat(user, span_warning("You do not need to use a billow at this moment, the forge is already hot enough!"))
-			in_use = FALSE
+			fail_message(user, "You cannot use [forge_item] without some sort of fuel in [src]!")
 			return
-		to_chat(user, span_warning("You start to pump the billow into the forge..."))
-		if(!do_after(user, forge_item.work_time, target = src)) //wait 3 seconds to upgrade (6 for primitive)
-			to_chat(user, span_warning("You abandon billowing the forge."))
-			in_use = FALSE
+		if(forge_temperature >= MAX_FORGE_TEMP) //we don't want the "temp" to overflow or something somehow
+			fail_message(user, "You do not need to use [forge_item] at this moment, [src] is already hot enough!")
```

Comment:
```suggestion
			fail_message(user, "You can't heat [src] to be any hotter!")
```
Better to give a strict "You CANT" instead of "You dont need to" for a failure, especially a limit-failure like this